src/phoenix/server/daemons/db_disk_usage_monitor.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+import logging
+from asyncio import sleep
+from datetime import datetime, timedelta, timezone
+from typing import Optional, cast
+
+import sqlalchemy as sa
+from email_validator import EmailNotValidError, validate_email
+from sqlalchemy import text
+from typing_extensions import assert_never
+
+from phoenix.config import (
+    get_env_database_allocated_storage_capacity_gibibytes,
+    get_env_database_usage_email_notification_threshold_percentage,
+    get_env_database_usage_insertion_blocking_threshold_percentage,
+)
+from phoenix.db import models
+from phoenix.db.helpers import SupportedSQLDialect
+from phoenix.server.email.types import DbUsageWarningEmailSender
+from phoenix.server.types import DaemonTask, DbSessionFactory
+
+logger = logging.getLogger(__name__)
+
+_SLEEP_SECONDS = 1
+_EMAIL_FREQUENCY_HOURS = 24
+_BYTES_PER_GIBIBYTE = 1024**3
+
+
+class DbDiskUsageMonitor(DaemonTask):
+    """
+    Monitors database disk space usage and triggers warnings/blocking when thresholds are exceeded.
+
+    This daemon:
+    - Periodically checks current database size
+    - Compares usage against configured thresholds
+    - Sends warning emails to admins when warning threshold is reached
+    - Toggles insertion blocking when blocking threshold is reached
+    """
+
+    def __init__(
+        self,
+        db: DbSessionFactory,
+        email_sender: Optional[DbUsageWarningEmailSender] = None,
+    ) -> None:
+        super().__init__()
+        self._db = db
+        self._email_sender = email_sender
+        # Tracks last email send time per admin email address to prevent spam
+        self._last_email_sent: dict[str, datetime] = {}
+
+    @property
+    def _is_disabled(self) -> bool:
+        return not bool(
+            get_env_database_allocated_storage_capacity_gibibytes()
+            and (
+                get_env_database_usage_email_notification_threshold_percentage()
+                or get_env_database_usage_insertion_blocking_threshold_percentage()
+            )
+        )
+
+    async def _run(self) -> None:
```

Comment:
💭 I wonder if we could export a prometheus counter for percentage. Would make monitoring easier.